File: Week Assignment 4 using C/3_reverse_dll.c
// 3. Reverse a Doubly Linked List
#include <stdio.h>
#include <stdlib.h>

struct Node { int data; struct Node* next; struct Node* prev; };

struct Node* push(struct Node* head, int data) {
    struct Node* n=(struct Node*)malloc(sizeof(struct Node));
    n->data=data; n->next=head; n->prev=NULL;
    if (head) head->prev = n;
    return n;
}

struct Node* reverseDLL(struct Node* head) {
    struct Node* cur = head;
    struct Node* temp = NULL;
    while (cur) {
        // swap next and prev
        temp = cur->prev;
        cur->prev = cur->next;
        cur->next = temp;
        head = cur; // update head to current
        cur = cur->prev; // moved forward using old next
    }
    return head;
}

void printList(struct Node* head) {
    printf("[");
    struct Node* t=head; int first=1;
    while (t) { if (!first) printf(", "); printf("%d", t->data); first=0; t=t->next; }
    printf("]\n");
}

int main() {
    struct Node* head = NULL;
    head = push(head, 3); head = push(head, 2); head = push(head, 1); // 1,2,3
    printList(head);
    head = reverseDLL(head); // 3,2,1
    printList(head);
    return 0;
}

File: Week Assignement 4 using Python/3_reverse_dll.py
# 3. Reverse a doubly linked list
class Node:
    def __init__(self,data):
        self.data=data; self.next=None; self.prev=None

def push(head,data):
    n=Node(data); n.next=head
    if head: head.prev = n
    return n

def reverse_dll(head):
    cur = head; temp = None; new_head = head
    while cur:
        temp = cur.prev
        cur.prev = cur.next
        cur.next = temp
        new_head = cur
        cur = cur.prev
    return new_head

def to_list(head):
    res=[]; cur=head
    while cur: res.append(cur.data); cur=cur.next
    return res

if __name__ == "__main__":
    head=None
    head = push(head,3); head=push(head,2); head=push(head,1) # 1,2,3
    print(to_list(head))
    head = reverse_dll(head)
    print(to_list(head))
